File: client/src/main/java/com/jraska/github/client/ui/RepoDetailActivity.java
package com.jraska.github.client.ui;

import android.app.Activity;
import android.arch.lifecycle.LiveData;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import butterknife.BindView;
import butterknife.OnClick;
import com.airbnb.epoxy.SimpleEpoxyAdapter;
import com.jraska.github.client.R;
import com.jraska.github.client.users.RepoDetail;
import com.jraska.github.client.users.RepoDetailViewModel;

public class RepoDetailActivity extends BaseActivity {
  private static final String EXTRA_FULL_REPO_NAME = "fullRepoName";

  @BindView(R.id.repo_detail_recycler) RecyclerView recyclerView;

  private RepoDetailViewModel viewModel;

  private String fullRepoName() {
    return getIntent().getStringExtra(EXTRA_FULL_REPO_NAME);
  }

  @Override protected void onCreate(@Nullable Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(R.layout.activity_repo_detail);
    recyclerView.setLayoutManager(new LinearLayoutManager(this));

    setTitle(fullRepoName());

    viewModel = viewModel(RepoDetailViewModel.class);
    LiveData<RepoDetailViewModel.ViewState> liveData = viewModel.repoDetail(fullRepoName());
    liveData.observe(this, this::setState);
  }

  @OnClick(R.id.repo_detail_github_fab) void onFitHubIconClicked(){
    viewModel.onFitHubIconClicked(fullRepoName());
  }

  void setState(RepoDetailViewModel.ViewState viewState) {
    if (viewState.repoDetail != null) {
      setRepoDetail(viewState.repoDetail);
    }

    if (viewState.error != null) {
      setError(viewState.error);
    }
  }

  void setError(Throwable error) {
    ErrorHandler.displayError(error, recyclerView);
  }

  void setRepoDetail(RepoDetail repoDetail) {
    SimpleEpoxyAdapter adapter = new SimpleEpoxyAdapter();
    adapter.addModels(new RepoDetailHeaderModel(repoDetail));

    String languageText = getString(R.string.repo_detail_language_used_template,
      repoDetail.data.language);
    adapter.addModels(new SimpleTextModel(languageText));

    String issuesText = getString(R.string.repo_detail_issues_template,
      String.valueOf(repoDetail.data.issuesCount));
    adapter.addModels(new SimpleTextModel(issuesText));

    recyclerView.setAdapter(adapter);
  }

  public static void start(Activity from, String fullRepoName) {
    Intent intent = new Intent(from, RepoDetailActivity.class);
    intent.putExtra(EXTRA_FULL_REPO_NAME, fullRepoName);

    from.startActivity(intent);
  }
}

File: client/src/main/java/com/jraska/github/client/ui/RepoDetailActivity.kt
package com.jraska.github.client.ui

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.airbnb.epoxy.SimpleEpoxyAdapter
import com.airbnb.epoxy.SimpleEpoxyModel
import com.jraska.github.client.R
import com.jraska.github.client.core.android.BaseActivity
import com.jraska.github.client.users.RepoDetail
import com.jraska.github.client.users.RepoDetailViewModel
import com.jraska.github.client.core.android.viewModel
import kotlinx.android.synthetic.main.activity_repo_detail.*
import kotlinx.android.synthetic.main.content_repo_detail.*

class RepoDetailActivity : BaseActivity() {

  private val viewModel: RepoDetailViewModel by lazy { viewModel(RepoDetailViewModel::class.java) }

  private fun fullRepoName(): String {
    return intent.getStringExtra(EXTRA_FULL_REPO_NAME)
  }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContentView(R.layout.activity_repo_detail)
    setSupportActionBar(toolbar)
    repo_detail_recycler.layoutManager = LinearLayoutManager(this)

    title = fullRepoName()

    val liveData = viewModel.repoDetail(fullRepoName())
    liveData.observe(this, Observer { this.setState(it) })

    repo_detail_github_fab.setOnClickListener { viewModel.onFitHubIconClicked(fullRepoName()) }
  }

  private fun setState(state: RepoDetailViewModel.ViewState) {
    when (state) {
      is RepoDetailViewModel.ViewState.Loading -> showLoading()
      is RepoDetailViewModel.ViewState.Error -> setError(state.error)
      is RepoDetailViewModel.ViewState.ShowRepo -> setRepoDetail(state.repo)
    }
  }

  private fun showLoading() {
    repo_detail_recycler.adapter = SimpleEpoxyAdapter().apply { addModels(SimpleEpoxyModel(R.layout.item_loading)) }
  }

  private fun setError(error: Throwable) {
    ErrorHandler.displayError(error, repo_detail_recycler)
  }

  private fun setRepoDetail(repoDetail: RepoDetail) {
    val adapter = SimpleEpoxyAdapter()
    adapter.addModels(RepoDetailHeaderModel(repoDetail))

    val languageText = getString(
      R.string.repo_detail_language_used_template,
      repoDetail.data.language
    )
    adapter.addModels(SimpleTextModel(languageText))

    val issuesText = getString(
      R.string.repo_detail_issues_template,
      repoDetail.data.issuesCount.toString()
    )
    adapter.addModels(SimpleTextModel(issuesText))

    repo_detail_recycler.adapter = adapter
  }

  companion object {
    private const val EXTRA_FULL_REPO_NAME = "fullRepoName"

    fun start(from: Activity, fullRepoName: String) {
      val intent = Intent(from, RepoDetailActivity::class.java)
      intent.putExtra(EXTRA_FULL_REPO_NAME, fullRepoName)

      from.startActivity(intent)
    }
  }
}
